### ISSUE TITLE
BUG - FlatList automatically scrolls to first when data is too large in horizontal mode RTL (Android)

### ISSUE THREAD

**Issue by matinzd** [facebook/react-native#26436]
- [x] Review the documentation: https://facebook.github.io/react-native

-  [x]  Search for existing issues: https://github.com/facebook/react-native/issues

- [x]  Use the latest React Native release: https://github.com/facebook/react-native/releases

## Environment
```
System:
    OS: macOS 10.15
    CPU: (6) x64 Intel(R) Core(TM) i5-8400 CPU @ 2.80GHz
    Memory: 448.96 MB / 8.00 GB
    Shell: 3.2.57 - /bin/bash
  Binaries:
    Node: 12.10.0 - /usr/local/bin/node
    Yarn: 1.19.0 - /usr/local/bin/yarn
    npm: 6.11.3 - /usr/local/bin/npm
  SDKs:
    iOS SDK:
      Platforms: iOS 13.1, DriverKit 19.0, macOS 10.15, tvOS 13.0, watchOS 6.0
  IDEs:
    Android Studio: 3.5 AI-191.8026.42.35.5791312
    Xcode: 11.1/11A1027 - /usr/bin/xcodebuild
  npmPackages:
    react: 16.8.6 => 16.8.6 
    react-native: 0.61.2 => 0.61.2
```

React Native version: **0.61.2**

## Code

```
<FlatList
    style={{
        flexDirection: 'row',
        marginHorizontal: -wp('3%'),
        zIndex: 40
    }}
    initialNumToRender={10}
    data={this.state.data}
    horizontal={true}
    key={item => item.groupNo}
    keyExtractor={(item, index) => index.toString()}
    showsHorizontalScrollIndicator={false}
    showsVerticalScrollIndicator={false}
    ref={ref => (this.flatList = ref)}
    renderItem={({ item }) => (
        <GroupAyeItem item={item} />
    )}
/>

```
## Steps To Reproduce

1. Added some data to my FlatList
2. Add horizontal={true} property to flatlist
3. Set `I18nManager.allowRTL(true)`
4. Set `I18nManager.forceRTL(true)`
5. Make my renderItem component extends as PureComponent.
6. Tested with Real Android Device and Emulator

As you see below when my app is running on RTL direction `FlatList` automatically scrolls to top/first of the list but LTR direction is OK! 

**RTL:**    

![RTL_GIF](https://user-images.githubusercontent.com/24797481/64906891-03c9b280-d701-11e9-8911-923f9035dc19.gif)

**LTR:**   

![LTR_GIF](https://user-images.githubusercontent.com/24797481/64906899-0e844780-d701-11e9-91cc-de5c4f7a6ef9.gif)

**UPDATE ON THIS ISSUE ! :**    
onScroll will get called rapidly in` VirtualizedList.js:1443` when RTL is enabled without scrolling!

![flatlist](https://user-images.githubusercontent.com/24797481/66717724-19f28d80-ede9-11e9-977b-bf78699bdc61.gif)

But in LTR :    
![flatlist_ltr](https://user-images.githubusercontent.com/24797481/66717774-b2890d80-ede9-11e9-892a-d3d5d95679e9.gif)

**Comment by matinzd** [comment]
Any update on this issue?!?

**Comment by matinzd** [comment]
> Hey there, it looks like there has been no activity on this issue recently. Has the issue been fixed, or does it still require the community's attention? This issue may be closed if no further activity occurs. You may also label this issue as a "Discussion" or add it to the "Backlog" and I will leave it open. Thank you for your contributions.

This issue does exist yet!

**Comment by matinzd** [comment]
label:disucssion

**Comment by djhr** [comment]
Having similar issue with scrollview on android, when size changes (e.g. due keyboard open), the scrollview resets it's position

**Comment by matinzd** [comment]
I had this problem even if there is no resize!

**Comment by mohs3n71** [comment]
I have the same exact problem too and couldn't find any fix for it

**Comment by vinayp-go** [comment]
Any Update on this? Or any alternative approach?

**Comment by mm7mod** [comment]
yes facing the same issue, I have found multiple issues for the same error ... the strange thing is that this issue has been around for years now without a solution .... 

any idea for a horizontal list views libraries out there that don't depend on the flat list ... I have found this library for vertical list: recyclerlistview and till now it is way way better than  the buggy flatlist and performance is much better

**Comment by fabriziobertoglio1987** [comment]
The issue should mention `RTL` in the title

Probably a duplicate of https://github.com/facebook/react-native/issues/19150

**Comment by praveens96** [comment]
even LTR has the issue in horizontal

**Comment by matinzd** [comment]
It has not fixed yet.

**Comment by kartavyaparekh96** [comment]
Any update?

**Comment by lucascooelho** [comment]
I've been looking for any solution about this issue for days so far. I haven't found.

**Comment by zohaib201** [comment]
You may follow this, https://stackoverflow.com/a/62066090/10464651
<FlatList
  ref={(ref) => { this.chatFlatList = ref; }}
  style={styles.flatList}
  data={this.state.items}
  renderItem={this._renderItem}
  maintainVisibleContentPosition={{
     minIndexForVisible: 0,
  }}
/>

**Comment by emil45** [comment]
Any update? facing the same issue with Hebrew (RTL language).
Maybe any workaround?

**Comment by lucascooelho** [comment]
> Any update? facing the same issue with Hebrew (RTL language).
> Maybe any workaround?

taka a look at this https://github.com/satya164/react-native-tab-view/issues/1068#issuecomment-783233062. It worked for me.

**Comment by matinzd** [comment]
> > Any update? facing the same issue with Hebrew (RTL language).
> > Maybe any workaround?
> 
> taka a look at this [satya164/react-native-tab-view#1068 (comment)](https://github.com/satya164/react-native-tab-view/issues/1068#issuecomment-783233062). It worked for me.

This is unrelated to the current issue.

**Comment by Noitham** [comment]
Facing the same issue here.. 
Tried both a FlatList & ScrollView, same issue.
https://user-images.githubusercontent.com/32934478/134426408-9b4af234-2f96-4cdf-a52d-55574d48c96d.mov

**Comment by aliraza-noon** [comment]
its really strange this is still a thing

**Comment by Chronostyle** [comment]
@matinzd hey, did you already find a solution for this bug?i also has the same issue just like you, for me it's from LTR not RTL on android...

**Comment by Kelvin-Lim-PKM** [comment]
I had the same issue. For my case it was caused by getItemLayout() in FlatList. Not using getItemLayout() solved the issue for me. Maybe a correct/precise implementation (calculation of length and offset) of getItemLayout() function is needed.

**Comment by matinzd** [comment]
> I had the same issue. For my case it was caused by getItemLayout() in FlatList. Not using getItemLayout() solved the issue for me. Maybe a correct/precise implementation (calculation of length and offset) of getItemLayout() function is needed.

But this was just happening on RTL. It should bahave the same way. Since then, I have changed 2 companies but this issue still exists 😅

**Comment by dev-arson** [comment]
https://github.com/facebook/react-native/pull/37651
In rn-tester app this fix works

**Comment by cjhines** [comment]
Experiencing this on horizontal FlatList, even without RTL.

After 10 items the list goes back to the beginning

**Comment by Kaveh-ap** [comment]
I'm facing the same problem!
I've tried FlashList, and unfortunately, it acts even weirder when RTL is true.

**Comment by BStar098** [comment]
Has this bug been fixed?

**Comment by emxxm1** [comment]
In my case, I was using a custom component both for horizontal & vertical flat list,
  <FlatList
                data={data}
                showsHorizontalScrollIndicator={false}
                pagingEnabled={true}
                horizontal={horizontal}
                keyExtractor={(item, index) => `key-${index}`}
                scrollEnabled={data?.length > 0}
                initialNumToRender={3}
                nestedScrollEnabled
                onEndReachedThreshold={0.8}
                initialScrollIndex={0}
                onEndReached={() => { }}
                showsVerticalScrollIndicator={false}
                renderItem={({ item }) => ...)}
            /> .
            
            This line  pagingEnabled={true} was a cause in my scenario, as paging is being true for horizontal cases too.
            I just changed it to  pagingEnabled={!horizontal}
            Maybe it will help any of you guys having the same scenario as me.

**Comment by tamighi** [comment]
I have the same issue for android only, RTL

**Comment by awais-wego** [comment]
I have the same issue for android only, RTL

**Comment by turker0** [comment]
Facing same issue, I found some workarounds which may be helpful 
- updating react native version (there are some rtl fix pr just merged)
- force flatlist to scroll to first index / element (you can also hide the component during this animtion for better ux using entering from re-animated)
- using getItemLayour for fixed width (for horizontal)
```
  getItemLayout={(data, index) => (
    {length: ITEM_WIDTH, offset: ITEM_WIDTH * index, index}
  )}
```

**Comment by hsaidc** [comment]
I experienced this issue after some styling updates. I reverted my latest updates one by one. When I removed the `flex: 1` in `contentContainerStyle` property, it is fixed. I hope this helps everyone experiencing the issue.

**Comment by aattalla** [comment]
Has any fix for this issue been released?

**Comment by dprajapati1179** [comment]
Facing same issue on iOS too, on first load it don't allow to scroll  unable to find the root cause

**Comment by aliraza-noon** [comment]
NOT STALE

**Comment by aliraza-noon** [comment]
not stale
